File: scripts/extract-epson-sii.py
"""
extract-epson-sii.py
Extract EPSON and SII movement data from the combined 55-page PDF.

Brand detection: products containing "SII" in text → SII; else → EPSON.
  SII series: VR, PC, VC, VD, VJ, VK prefixes — text explicitly says "XXX SII"
  EPSON series: VX, VH, Y, YM — text says "Epson XXX" (lowercase)

Outputs:
  src/data/movimientos-epson.json
  src/data/movimientos-sii.json
  public/images/productos/movimientos-epson/<safe>.png
  public/images/productos/movimientos-sii/<safe>.png
  public/catalogo-pages/movimientos-epson/page-XX.jpg  (sequential)
  public/catalogo-pages/movimientos-sii/page-XX.jpg    (sequential)
"""

import fitz
import re
import json
import os
from PIL import Image
import numpy as np

# ─── Config ─────────────────────────────────────────────────────────────────

PDF_PATH   = "public/catalogos/movimientos-epson.pdf"
EPSON_IMG  = "public/images/productos/movimientos-epson"
SII_IMG    = "public/images/productos/movimientos-sii"
EPSON_CATS = "public/catalogo-pages/movimientos-epson"
SII_CATS   = "public/catalogo-pages/movimientos-sii"
EPSON_JSON = "src/data/movimientos-epson.json"
SII_JSON   = "src/data/movimientos-sii.json"

TARGET_H   = 270
GAP        = 20

for d in [EPSON_IMG, SII_IMG, EPSON_CATS, SII_CATS]:
    os.makedirs(d, exist_ok=True)

# ─── Image helpers ───────────────────────────────────────────────────────────

def safe(modelo):
    return modelo.replace('/', '_').replace(' ', '_')

def render_rect(page, rect, scale=2.5):
    mat = fitz.Matrix(scale, scale)
    pix = page.get_pixmap(matrix=mat, clip=rect, colorspace=fitz.csRGB)
    return np.frombuffer(pix.samples, dtype=np.uint8).reshape(pix.height, pix.width, 3).copy()

def render_full_page(page, scale=1.5):
    mat = fitz.Matrix(scale, scale)
    pix = page.get_pixmap(matrix=mat, colorspace=fitz.csRGB)
    return np.frombuffer(pix.samples, dtype=np.uint8).reshape(pix.height, pix.width, 3).copy()

def crop_white(arr, threshold=245, pad=8):
    dark = (arr[:, :, 0] < threshold) | (arr[:, :, 1] < threshold) | (arr[:, :, 2] < threshold)
    rows = np.any(dark, axis=1)
    cols = np.any(dark, axis=0)
    if not np.any(rows):
        return arr
    r0 = max(0, int(np.argmax(rows)) - pad)
    r1 = min(arr.shape[0], arr.shape[0] - int(np.argmax(rows[::-1])) + pad)
    c0 = max(0, int(np.argmax(cols)) - pad)
    c1 = min(arr.shape[1], arr.shape[1] - int(np.argmax(cols[::-1])) + pad)
    return arr[r0:r1, c0:c1]

def resize_to_height(arr, h):
    ih, iw = arr.shape[:2]
    if ih == 0 or iw == 0:
        return arr
    nw = max(1, int(iw * h / ih))
    return np.array(Image.fromarray(arr).resize((nw, h), Image.LANCZOS))

def stitch(crops):
    scaled = [resize_to_height(c, TARGET_H) for c in crops]
    white_gap = np.full((TARGET_H, GAP, 3), 255, dtype=np.uint8)
    parts = []
    for i, s in enumerate(scaled):
        if i > 0:
            parts.append(white_gap)
        parts.append(s)
    result = np.concatenate(parts, axis=1)
    pad = 16
    out = np.full((TARGET_H + pad * 2, result.shape[1] + pad * 2, 3), 255, dtype=np.uint8)
    out[pad:pad + TARGET_H, pad:pad + result.shape[1]] = result
    return out

def extract_product_png(page, img_dir, safe_name):
    imgs = page.get_images(full=True)
    if not imgs:
        return None
    rects_data = []
    for img_info in imgs:
        xref = img_info[0]
        img_rects = page.get_image_rects(xref)
        if img_rects:
            rects_data.append((img_rects[0].y0, img_rects[0], xref))
    if not rects_data:
        return None
    rects_data.sort(key=lambda t: t[0])
    crops = []
    for _, rect, _ in rects_data:
        rendered = render_rect(page, rect, scale=2.5)
        cropped  = crop_white(rendered)
        if cropped.shape[0] > 10 and cropped.shape[1] > 10:
            crops.append(cropped)
    if not crops:
        return None
    if len(crops) == 1:
        final = resize_to_height(crops[0], TARGET_H)
        pad = 16
        out = np.full((TARGET_H + pad * 2, final.shape[1] + pad * 2, 3), 255, dtype=np.uint8)
        out[pad:pad + TARGET_H, pad:pad + final.shape[1]] = final
        final = out
    else:
        print(f"    → {len(crops)} views → stitching")
        final = stitch(crops)
    out_path = os.path.join(img_dir, f"{safe_name}.png")
    Image.fromarray(final).save(out_path)
    return out_path

# ─── Text parser ─────────────────────────────────────────────────────────────

def parse_page(text, modelo):
    lc = text.lower()

    # brand — SII models always contain "SII" literally; EPSON models use "Epson" (lowercase)
    brand = 'SII' if 'SII' in text else 'EPSON'

    # tipo
    is_crono = bool(re.search(r'cron[oó]grafo|chrono', lc))
    is_dama  = bool(re.search(r'para damas?|movimiento de damas?|ladies?', lc)) and not is_crono
    tipo     = 'CRONÓGRAFO' if is_crono else ('DAMA' if is_dama else 'CABALLERO')

    # fecha
    fecha = bool(re.search(r'\bfecha\b|calendar|date', lc))

    # dimensiones — rectangular first, then round
    rect_m = re.search(
        r'dimensiones son\s*([\d,]+)\s*mm\s*[xX×]\s*([\d,]+)\s*mm|'
        r'dimensiones de\s*([\d,]+)\s*mm\s*[xX×]\s*([\d,]+)\s*mm|'
        r'unas dimensiones de\s*([\d,]+)\s*mm\s*[xX×]\s*([\d,]+)\s*mm|'
        r'([\d,]+)\s*mm\s*[xX×]\s*([\d,]+)\s*mm',
        text, re.I
    )
    dimensiones = ''
    if rect_m:
        groups = [g for g in rect_m.groups() if g is not None]
        a, b = groups[0], groups[1]
        dimensiones = f"{a.replace(',', '.')}mm × {b.replace(',', '.')}mm"
    else:
        # Round: "XX,XX mm (diámetro)" or "diámetro XX,XX mm" or "(XX,XX mm)" after "líneas"
        diam_m = re.search(r'([\d,]+)\s*mm\s*\(di[aá]metro\)', text, re.I)
        if not diam_m:
            diam_m = re.search(r'di[aá]metro\s+([\d,]+)\s*mm', text, re.I)
        if not diam_m:
            diam_m = re.search(r'l[ií]neas\s*\((\d+[,.]?\d*)\s*mm\)', text, re.I)
        if diam_m:
            dimensiones = f"⌀ {diam_m.group(1).replace(',', '.')}mm"

    # altura — multiple patterns in order of specificity
    alt_m = re.search(r'grosor(?:\s+es)?\s+de\s*([\d,]+)\s*mm', text, re.I)
    if not alt_m:
        alt_m = re.search(r'su grosor es\s+([\d,]+)\s*mm', text, re.I)
    if not alt_m:
        alt_m = re.search(r'altura total\s*[\w\s]*?([\d,]+)\s*mm', text, re.I)
    if not alt_m:
        alt_m = re.search(r'con una altura de\s*([\d,]+)\s*mm', text, re.I)
    if not alt_m:
        alt_m = re.search(r'altura de movimiento de\s*([\d,]+)\s*mm', text, re.I)
    if not alt_m:
        alt_m = re.search(r'altura de\s+([\d,]+)\s*mm', text, re.I)
    altura = alt_m.group(1).replace(',', '.') if alt_m else ''

    # joyas — handles "joya", "joyas", "rubíes", "rubí"
    joyas_m = re.search(r'(\d+)\s+(?:joyas?|rub[ií]es?)', text, re.I)
    joyas = int(joyas_m.group(1)) if joyas_m else 0

    # bateria — cascading patterns: "batería 371", "celda 371", "celular 371", "pila 364", "SR521SW"
    bat_m = re.search(r'bater[ií]a\s+(\d{3,4})', text, re.I)
    if not bat_m:
        bat_m = re.search(r'\b(?:celda|celular?|pila)\s+(\d{3,4})', text, re.I)
    if not bat_m:
        # SR code → map to numeric
        sr_m = re.search(r'\bSR(\d{3,4}[A-Z]*)\b', text, re.I)
        bateria = f"SR{sr_m.group(1)}" if sr_m else ''
    else:
        bateria = bat_m.group(1)

    # intercambios
    intc_m = re.search(
        r'(?:Movimiento Intercambio|Intercambio de movimiento)\s*([\s\S]+?)$',
        text, re.I
    )
    intercambios = intc_m.group(1).strip().replace('\n', ' ').replace('  ', ' ') if intc_m else ''

    return {
        "modelo":       modelo,
        "brand":        brand,
        "tipo":         tipo,
        "fecha":        fecha,
        "dimensiones":  dimensiones,
        "altura":       altura,
        "joyas":        joyas,
        "bateria":      bateria,
        "intercambios": intercambios,
    }

# ─── Scan all pages ───────────────────────────────────────────────────────────

doc = fitz.open(PDF_PATH)
print(f"PDF: {len(doc)} pages\n")

epson_pages = []
sii_pages   = []

for i in range(len(doc)):
    page = doc[i]
    text = page.get_text()
    if 'MODELO:' not in text:
        continue
    m = re.search(r'MODELO:\s*(\S+)', text)
    if not m:
        continue
    modelo = m.group(1).strip()
    brand  = 'SII' if 'SII' in text else 'EPSON'
    if brand == 'EPSON':
        epson_pages.append((i, modelo, text))
    else:
        sii_pages.append((i, modelo, text))

print(f"EPSON products: {len(epson_pages)}")
print(f"SII   products: {len(sii_pages)}")
print(f"EPSON: {[m for _, m, _ in epson_pages]}")
print(f"SII:   {[m for _, m, _ in sii_pages]}")

# ─── Process each brand ───────────────────────────────────────────────────────

BRANDS = [
    ("EPSON", epson_pages, EPSON_IMG, EPSON_CATS, EPSON_JSON, "movimientos-epson"),
    ("SII",   sii_pages,   SII_IMG,   SII_CATS,   SII_JSON,   "movimientos-sii"),
]

for brand_name, pages_list, img_dir, cat_dir, json_path, slug in BRANDS:
    print(f"\n{'='*60}")
    print(f"Processing {brand_name} — {len(pages_list)} products")
    print(f"{'='*60}")

    products = []

    for seq_idx, (pg_idx, modelo, text) in enumerate(pages_list):
        print(f"\n  [{seq_idx+1:2d}/{len(pages_list)}] {modelo} (PDF pg {pg_idx+1})")
        page = doc[pg_idx]

        data = parse_page(text, modelo)
        safe_name = safe(modelo)
        data["img"] = f"/images/productos/{slug}/{safe_name}.png"
        products.append(data)

        print(f"    {data['brand']} | {data['tipo']} | fecha={data['fecha']} | "
              f"dim={data['dimensiones']} | alt={data['altura']}mm | "
              f"bat={data['bateria']} | joyas={data['joyas']}")

        # Product PNG
        out_png = extract_product_png(page, img_dir, safe_name)
        if out_png:
            img_check = Image.open(out_png)
            print(f"    PNG: {img_check.width}×{img_check.height} → {out_png}")
        else:
            print(f"    PNG: FAILED")

        # Catalog page JPG (sequential numbering: seq_idx+2)
        cat_num  = seq_idx + 2
        cat_path = os.path.join(cat_dir, f"page-{str(cat_num).zfill(2)}.jpg")
        Image.fromarray(render_full_page(page)).save(cat_path, "JPEG", quality=85)

    # Cover page
    Image.fromarray(render_full_page(doc[0])).save(
        os.path.join(cat_dir, "page-01.jpg"), "JPEG", quality=85
    )
    print(f"\n  page-01.jpg (cover) saved")

    # JSON
    with open(json_path, 'w', encoding='utf-8') as f:
        json.dump(products, f, ensure_ascii=False, indent=2)
    print(f"  JSON → {json_path} ({len(products)} products)")

doc.close()

# ─── Summary ─────────────────────────────────────────────────────────────────

print("\n\n─── FIELD COVERAGE ───────────────────────────────────────────────")
for json_path, label in [(EPSON_JSON, 'EPSON'), (SII_JSON, 'SII')]:
    with open(json_path) as f:
        prods = json.load(f)
    print(f"\n{label} ({len(prods)} products):")
    for field in ['dimensiones', 'altura', 'bateria']:
        filled = sum(1 for p in prods if p.get(field))
        print(f"  {field:14s}: {filled}/{len(prods)}")

print("\n✅ Done.")

File: scripts/extract-epson-sii.mjs
/**
 * extract-epson-sii.mjs
 * Extracts product images from movimientos-epson.pdf.
 *
 * Two strategies:
 *  A) Page-render (PyMuPDF clip) — for pages with multiple separate embedded
 *     images (SMask/PNG overlay + photo), already stored in different rects.
 *  B) Raw-JPEG + vertical split — for FORCE_SPLIT models whose single embedded
 *     JPEG is a portrait composite (two views stacked top/bottom). We extract
 *     the raw bytes, detect the white horizontal band between views, crop each
 *     segment, then stitch side-by-side.
 *
 * All outputs get sharp's .flatten({ background: white }) for a guaranteed
 * pure-white background.
 *
 * Usage:  node scripts/extract-epson-sii.mjs
 */

import sharp from 'sharp';
import { execFileSync } from 'child_process';
import { writeFileSync, mkdirSync } from 'fs';
import { tmpdir } from 'os';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT      = path.resolve(__dirname, '..');

const EPSON_OUT = path.join(ROOT, 'public/images/productos/movimientos-epson');
const SII_OUT   = path.join(ROOT, 'public/images/productos/movimientos-sii');
const TARGET_H  = 302;
const GAP       = 20;
const PAD       = 16;
const WHITE     = { r: 255, g: 255, b: 255 };

// Models with a single portrait JPEG containing two stacked views.
// These receive the raw-bytes + vertical-split treatment.
const FORCE_SPLIT = new Set([
    // EPSON
    'VX00', 'VX01', 'VX3J', 'VX3KE', 'VX10', 'VX11', 'VX12-3/6',
    'VX32-3/6', 'VX33-3', 'VX42-3/6', 'VX43-3', 'VX50', 'VX51',
    'VX82-3/6', 'VX83-3', 'Y121', 'YM62-3', 'YM92-3',
    // SII
    'VR32-4', 'PC32-3/6', 'VC00', 'VD53', 'VD54', 'VD57',
]);

mkdirSync(EPSON_OUT, { recursive: true });
mkdirSync(SII_OUT,   { recursive: true });

// ─── Python renderer ──────────────────────────────────────────────────────────
// For each product page returns:
//   modelo, brand,
//   raw_b64  — raw bytes of the FIRST embedded image (strategy B)
//   crops_b64 — page-rendered crops per image rect (strategy A)
const PY_CODE = String.raw`
import fitz, json, base64, re, io, sys
import numpy as np
from PIL import Image

PDF   = 'public/catalogos/movimientos-epson.pdf'
SCALE = 3.0
PAD   = 14
THRESH = 238

def crop_content(arr):
    dark = (arr[:,:,0] < THRESH) | (arr[:,:,1] < THRESH) | (arr[:,:,2] < THRESH)
    rows = np.any(dark, axis=1)
    cols = np.any(dark, axis=0)
    if not np.any(rows):
        return arr
    r0 = max(0, int(np.argmax(rows)) - PAD)
    r1 = min(arr.shape[0], arr.shape[0] - int(np.argmax(rows[::-1])) + PAD)
    c0 = max(0, int(np.argmax(cols)) - PAD)
    c1 = min(arr.shape[1], arr.shape[1] - int(np.argmax(cols[::-1])) + PAD)
    return arr[r0:r1, c0:c1]

doc    = fitz.open(PDF)
result = []

for i in range(len(doc)):
    page = doc[i]
    text = page.get_text()
    if 'MODELO:' not in text:
        continue
    m = re.search(r'MODELO:\s*(\S+)', text)
    if not m:
        continue
    modelo = m.group(1).strip()
    brand  = 'SII' if 'SII' in text else 'EPSON'

    imgs = page.get_images(full=True)

    # Raw bytes of first embedded image (for FORCE_SPLIT strategy B)
    raw_b64 = ''
    if imgs:
        xref    = imgs[0][0]
        img_raw = doc.extract_image(xref)['image']
        raw_b64 = base64.b64encode(img_raw).decode()

    # Page-rendered crops per rect (for strategy A — multi-image pages)
    rects_data = []
    for img_info in imgs:
        xref      = img_info[0]
        img_rects = page.get_image_rects(xref)
        if img_rects:
            r = img_rects[0]
            rects_data.append((float(r.y0), r))
    rects_data.sort(key=lambda t: t[0])

    crops_b64 = []
    for _, rect in rects_data:
        mat = fitz.Matrix(SCALE, SCALE)
        pix = page.get_pixmap(matrix=mat, clip=rect, colorspace=fitz.csRGB)
        arr = np.frombuffer(pix.samples, dtype=np.uint8).reshape(pix.height, pix.width, 3).copy()
        cropped = crop_content(arr)
        if cropped.shape[0] < 20 or cropped.shape[1] < 20:
            cropped = arr
        buf = io.BytesIO()
        Image.fromarray(cropped).save(buf, format='PNG')
        crops_b64.append(base64.b64encode(buf.getvalue()).decode())

    result.append({
        'modelo':    modelo,
        'brand':     brand,
        'raw_b64':   raw_b64,
        'crops_b64': crops_b64,
    })

doc.close()
json.dump(result, sys.stdout, ensure_ascii=False)
`;

const tmpPy = path.join(tmpdir(), '_render_epson_sii.py');
writeFileSync(tmpPy, PY_CODE, 'utf8');

console.log('Rendering PDF pages via PyMuPDF…');
const jsonStr = execFileSync('python3', [tmpPy], {
    cwd: ROOT,
    maxBuffer: 500 * 1024 * 1024,
}).toString();
const pages = JSON.parse(jsonStr);
console.log(`Parsed ${pages.length} products\n`);

// ─── Sharp helpers ────────────────────────────────────────────────────────────

/**
 * Returns { top, bot, left, right } pixel bounds of non-white content.
 * "White" = all channels >= thresh.
 */
async function contentBounds(buf, thresh = 238) {
    const { data, info } = await sharp(buf)
        .flatten({ background: WHITE })
        .raw()
        .toBuffer({ resolveWithObject: true });
    const { width, height, channels } = info;
    let top = height, bot = -1, left = width, right = -1;
    for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
            const i = (y * width + x) * channels;
            if (data[i] < thresh || data[i + 1] < thresh || data[i + 2] < thresh) {
                if (y < top)   top   = y;
                if (y > bot)   bot   = y;
                if (x < left)  left  = x;
                if (x > right) right = x;
            }
        }
    }
    return { top, bot, left, right, width, height };
}

/**
 * Splits a portrait image (two views stacked top/bottom, separated by a white
 * horizontal band) into an array of PNG buffers — one per segment.
 * Returns the original [buf] unchanged if no clean split is found.
 */
async function splitVertical(buf) {
    const { data, info } = await sharp(buf)
        .flatten({ background: WHITE })
        .raw()
        .toBuffer({ resolveWithObject: true });
    const { width, height, channels } = info;

    // Which rows are "all white"?
    const isWhite = new Uint8Array(height);
    for (let y = 0; y < height; y++) {
        let w = true;
        for (let x = 0; x < width && w; x++) {
            const i = (y * width + x) * channels;
            if (data[i] < 245 || data[i + 1] < 245 || data[i + 2] < 245) w = false;
        }
        isWhite[y] = w ? 1 : 0;
    }

    // Find content segments (runs of non-white rows)
    const segs = [];
    let inSeg = false, start = 0;
    for (let y = 0; y < height; y++) {
        if (!isWhite[y] && !inSeg) { start = y; inSeg = true; }
        else if (isWhite[y] && inSeg) { segs.push([start, y]); inSeg = false; }
    }
    if (inSeg) segs.push([start, height]);

    // Need at least 2 meaningful segments (each >= 40px tall)
    const bigSegs = segs.filter(([a, b]) => b - a >= 40);
    if (bigSegs.length < 2) return [buf];

    // Crop each segment with padding, return as PNG buffers
    const SEG_PAD = 8;
    const crops = [];
    for (const [segTop, segBot] of bigSegs) {
        const top = Math.max(0, segTop - SEG_PAD);
        const bot = Math.min(height, segBot + SEG_PAD);
        const segBuf = await sharp(buf)
            .flatten({ background: WHITE })
            .extract({ left: 0, top, width, height: bot - top })
            .png()
            .toBuffer();
        crops.push(segBuf);
    }
    return crops;
}

/**
 * Crop whitespace from all sides of a buffer, then resize to TARGET_H.
 * Returns { buf: PNG, width, height }.
 */
async function cropAndResize(inputBuf) {
    const b = await contentBounds(inputBuf);
    if (b.bot < 0) return null; // blank

    const SEG_PAD = 10;
    const left   = Math.max(0, b.left  - SEG_PAD);
    const top    = Math.max(0, b.top   - SEG_PAD);
    const right  = Math.min(b.width,  b.right  + SEG_PAD + 1);
    const bot    = Math.min(b.height, b.bot    + SEG_PAD + 1);
    const cropW  = right - left;
    const cropH  = bot   - top;
    const newW   = Math.max(1, Math.round(cropW * TARGET_H / cropH));

    const buf = await sharp(inputBuf)
        .flatten({ background: WHITE })
        .extract({ left, top, width: cropW, height: cropH })
        .resize(newW, TARGET_H, { fit: 'fill' })
        .png()
        .toBuffer();

    return { buf, width: newW, height: TARGET_H };
}

/** Fraction of pixels darker than 200 brightness. */
async function darkFrac(inputBuf) {
    const { data, info } = await sharp(inputBuf)
        .flatten({ background: WHITE })
        .raw()
        .toBuffer({ resolveWithObject: true });
    const n = info.width * info.height;
    let dark = 0;
    for (let i = 0; i < n * 3; i += 3) {
        if ((data[i] + data[i + 1] + data[i + 2]) / 3 < 200) dark++;
    }
    return dark / n;
}

/** Horizontal stitch: GAP between crops, PAD around the whole canvas. */
async function stitchH(crops) {
    const totalW = crops.reduce((s, c) => s + c.width, 0)
                 + GAP * (crops.length - 1) + PAD * 2;
    const totalH = TARGET_H + PAD * 2;
    let x = PAD;
    const composites = crops.map(({ buf, width }) => {
        const comp = { input: buf, left: x, top: PAD };
        x += width + GAP;
        return comp;
    });
    return sharp({ create: { width: totalW, height: totalH, channels: 3, background: WHITE } })
        .composite(composites)
        .flatten({ background: WHITE })
        .png()
        .toBuffer();
}

/** Single crop with PAD canvas. */
async function single(crop) {
    const { buf, width, height } = crop;
    return sharp({ create: { width: width + PAD * 2, height: height + PAD * 2, channels: 3, background: WHITE } })
        .composite([{ input: buf, left: PAD, top: PAD }])
        .flatten({ background: WHITE })
        .png()
        .toBuffer();
}

// ─── Main loop ────────────────────────────────────────────────────────────────
for (const { modelo, brand, raw_b64, crops_b64 } of pages) {
    const safe    = modelo.replace(/\//g, '_');
    const outDir  = brand === 'SII' ? SII_OUT : EPSON_OUT;
    const outPath = path.join(outDir, `${safe}.png`);

    let rawSegBufs = [];   // PNG buffers for each independent view
    let strategy   = '';

    if (FORCE_SPLIT.has(modelo) && raw_b64) {
        // ── Strategy B: raw JPEG → detect vertical segments → split ──
        const rawBuf = Buffer.from(raw_b64, 'base64');
        rawSegBufs   = await splitVertical(rawBuf);
        strategy     = rawSegBufs.length > 1
            ? `raw-JPEG split → ${rawSegBufs.length} segs`
            : 'raw-JPEG (no split found)';
    } else {
        // ── Strategy A: page-rendered crops per image rect ──
        rawSegBufs = (crops_b64 ?? []).map(b => Buffer.from(b, 'base64'));
        strategy   = `page-render × ${rawSegBufs.length}`;
    }

    if (rawSegBufs.length === 0) {
        console.log(`  [SKIP] ${brand} ${modelo} — no image data`);
        continue;
    }

    // Crop whitespace from each segment + resize to TARGET_H
    const resized = (await Promise.all(rawSegBufs.map(cropAndResize)))
        .filter(Boolean);

    // Drop completely blank views (>99.5% white)
    const valid = [];
    for (const r of resized) {
        if ((await darkFrac(r.buf)) > 0.005) valid.push(r);
    }
    const final = valid.length > 0 ? valid : resized.slice(0, 1);

    if (final.length === 0) {
        console.log(`  [SKIP] ${brand} ${modelo} — all blank`);
        continue;
    }

    const outBuf = final.length === 1
        ? await single(final[0])
        : await stitchH(final);

    await sharp(outBuf).toFile(outPath);
    const meta = await sharp(outPath).metadata();
    console.log(`  ${brand} ${modelo}: ${meta.width}×${meta.height}  [${strategy}]`);
}

console.log('\n✅ Done. All images overwritten.');
